File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/listview/DataViewController.java
package com.goexp.galgame.gui.view.game.listview;

import com.goexp.galgame.common.model.GameState;
import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.task.PanelTask;
import com.goexp.galgame.gui.task.TaskService;
import com.goexp.galgame.gui.view.DefaultController;
import com.goexp.galgame.gui.view.game.listview.sidebar.BrandGroupController;
import com.goexp.galgame.gui.view.game.listview.sidebar.DateGroupController;
import com.goexp.galgame.gui.view.game.listview.sidebar.FilterPanelController;
import com.goexp.galgame.gui.view.game.listview.simplelist.SimpleListViewController;
import com.goexp.galgame.gui.view.game.listview.simplelist.small.ListViewController;
import com.goexp.galgame.gui.view.game.listview.tableview.TableController;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.collections.ObservableList;
import javafx.collections.transformation.FilteredList;
import javafx.collections.transformation.SortedList;
import javafx.concurrent.Service;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Region;
import javafx.scene.layout.VBox;

import java.util.List;
import java.util.function.Predicate;

public class DataViewController extends DefaultController {

    public BooleanProperty reloadProperty = new SimpleBooleanProperty(false);

    /**
     * Controllers
     */
    @FXML
    public TableController tableViewController;
    @FXML
    private FilterPanelController filterPanelController;
    @FXML
    private BrandGroupController brandGroupController;
    @FXML
    private DateGroupController dateGroupController;


    /**
     * Status bar
     */
    @FXML
    private Label lbItemCount;
    @FXML
    public ProgressBar progessloading;


    /**
     * Toggle
     */
    @FXML
    private ToggleGroup gameViewChange;
    @FXML
    private ToggleButton toggList;
    @FXML
    private ToggleButton toggGrid;
    @FXML
    private ToggleButton toggSmall;
    @FXML
    private ToggleButton toggImg;


    /**
     * main panel
     */
    @FXML
    private TableView<Game> tableView;
    @FXML
    private ListView<Game> listSimple;
//    @FXML
//    private ListView<Game> smallListSimple;
    @FXML
    private Region imgView;


    /**
     * Sidebar
     */
    @FXML
    public Accordion groupPanel;
    @FXML
    private Region filterPanel;
    @FXML
    private Button btnHide;
    @FXML
    private VBox tagFlow;


    private FilteredList<Game> filteredGames;

    private Predicate<Game> groupPredicate;

    private Service<List<HBox>> groupTagServ = new TaskService<>(() -> new PanelTask.GroupTag(filteredGames));


    protected void initialize() {
        initSwitchBar();
        initSideBar();
        initTagPanel();
    }

    private void initTagPanel() {
        groupTagServ.valueProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue != null) {
                tagFlow.getChildren().setAll(newValue);
            }
        });
    }

    private void initSwitchBar() {
        toggGrid.setUserData(tableView);
        toggSmall.setUserData(listSimple);
        toggList.setUserData(listSimple);
        toggImg.setUserData(imgView);

        gameViewChange.selectedToggleProperty().addListener((observable, oldValue, checkedBox) -> {
            if (checkedBox != null) {
                ((Region) checkedBox.getUserData()).toFront();

                if (checkedBox == toggList) {
                    listSimple.setCellFactory(SimpleListViewController.cellFactory);

                } else if (checkedBox == toggSmall) {
                    listSimple.setCellFactory(ListViewController.cellFactory);
                }
            } else {
                gameViewChange.selectToggle(oldValue);
            }

        });
    }

    private void initSideBar() {
        btnHide.setUserData(true);
        btnHide.getProperties().put("groupPanel", groupPanel.getPrefWidth());
        btnHide.setOnAction(event -> {
            var state = (Boolean) btnHide.getUserData();
            var width = (double) btnHide.getProperties().get("groupPanel");

            state = !state;

            groupPanel.setVisible(state);
            groupPanel.setPrefWidth(state ? width : 0);
            btnHide.setUserData(state);

        });


        // #2
        groupPanel.setExpandedPane(groupPanel.getPanes().get(1));

        initSidebarContentView();
    }

    private void initSidebarContentView() {
        filterPanelController.onSetProperty.addListener((o, old, newV) -> {
            if (newV != null && newV) {
                var filterPredicate = filterPanelController.predicate;
                var p = groupPredicate != null ? filterPredicate.and(groupPredicate) : filterPredicate;
                filteredGames.setPredicate(p);
            }
        });

        brandGroupController.onSetProperty.addListener((observable, oldValue, newValue) -> {
            if (newValue) {
                groupPredicate = brandGroupController.predicate;

                var filterPredicate = filterPanelController.predicate;
                var p = filterPredicate != null ? groupPredicate.and(filterPredicate) : groupPredicate;
                filteredGames.setPredicate(p);
            }
        });
        dateGroupController.onSetProperty.addListener((observable, oldValue, newValue) -> {
            if (newValue) {
                groupPredicate = dateGroupController.predicate;

                var filterPredicate = filterPanelController.predicate;
                var p = filterPredicate != null ? groupPredicate.and(filterPredicate) : groupPredicate;
                filteredGames.setPredicate(p);
            }
        });
    }


    public void load(ObservableList<Game> games, Predicate<Game> initPredicate) {

        filterPanel.setVisible(true);

        groupPredicate = null;

        filteredGames = new FilteredList<>(games);
        filteredGames.predicateProperty().addListener((o, old, newV) -> {
            if (newV != null) {
                tableView.scrollTo(0);
                listSimple.scrollTo(0);
                resetCount(filteredGames);
            }
        });

        // set filter
        filteredGames.setPredicate(initPredicate);

        // set defaultPredicate
        filterPanelController.predicate = initPredicate;

        var sortedData = new SortedList<>(filteredGames);
        sortedData.comparatorProperty().bind(tableView.comparatorProperty());

        loadItems(sortedData);

        setSideBarData(filteredGames);
    }


    public void load(ObservableList<Game> games) {
        Predicate<Game> defaultP = g -> g.state.get().getValue() > GameState.BLOCK.getValue() && !(g.star > 0 && g.star < 3);
        load(games, defaultP);
    }

    private void setSideBarData(FilteredList<Game> filteredGames) {
        dateGroupController.init(filteredGames);
        brandGroupController.init(filteredGames);

        groupTagServ.restart();
    }

    private void loadItems(SortedList<Game> sortedData) {

        tableView.setItems(sortedData);
        tableView.scrollTo(0);

        listSimple.setCellFactory(ListViewController.cellFactory);

        listSimple.setItems(sortedData);
        listSimple.scrollTo(0);
    }

//    private void loadItems(ObservableList<Game> sortedData) {
//        resetCount(sortedData);
//        tableView.setItems(sortedData);
//        tableView.scrollTo(0);
//
//        listSimple.setItems(sortedData);
//        listSimple.scrollTo(0);
//    }

    private void resetCount(List<Game> filteredGames) {
        lbItemCount.setText(String.format("%d 件", filteredGames.size()));
    }


    @FXML
    private void reload_OnAction(ActionEvent event) {

        reloadProperty.set(true);
        reloadProperty.set(false);

    }
}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/listview/DataViewController.scala
package com.goexp.galgame.gui.view.game.listview

import com.goexp.galgame.gui.HGameApp
import com.goexp.galgame.gui.model.Game
import com.goexp.galgame.gui.task.game.panel.group.node.{DataItem, SampleItem}
import com.goexp.galgame.gui.task.game.panel.group.{ByCV, ByTag}
import com.goexp.galgame.gui.util.{SimpleFxmlLoader, Tags}
import com.goexp.galgame.gui.view.game.listview.sidebar.{BrandGroupController, DateGroupController, FilterPanelController}
import com.goexp.galgame.gui.view.game.listview.simplelist.small.HeaderController
import com.goexp.galgame.gui.view.game.listview.tablelist.TableListController
import com.goexp.ui.javafx.control.cell.NodeListCell
import com.goexp.ui.javafx.{DefaultController, TaskService}
import javafx.collections.transformation.{FilteredList, SortedList}
import javafx.collections.{FXCollections, ObservableList}
import javafx.fxml.FXML
import javafx.scene.control._
import javafx.scene.layout.{HBox, Region}
import org.controlsfx.control.{GridCell, GridView, PopOver}

import java.util
import java.util.function.Predicate
import scala.jdk.CollectionConverters._

class DataViewController extends DefaultController {
  /**
   * Controllers
   */
  @FXML var tablelistController: TableListController = _
  @FXML private var filterPanelController: FilterPanelController = _
  @FXML private var brandGroupController: BrandGroupController = _
  @FXML private var dateGroupController: DateGroupController = _
  /**
   * Status bar
   */
  @FXML private var lbItemCount: Label = _
  @FXML var loadingBar: ProgressBar = _
  /**
   * main panel
   */
  @FXML private var tablelist: TableView[Game] = _
  //  @FXML private var mainTab: TabPane = _
  /**
   * Sidebar
   */
  @FXML private var filterPanel: Region = _
  @FXML private var cvList: ListView[DataItem] = _
  @FXML private var tagList: ListView[DataItem] = _

  @FXML private var gridView: GridView[Game] = _
  private val popPanel = new PopOver

  private var filteredGames: FilteredList[Game] = _
  private var groupPredicate: Predicate[Game] = _

  final private val groupCVServ = TaskService(new ByCV(filteredGames))
  final private val groupTagServ = TaskService(new ByTag(filteredGames))

  override protected def initialize() = {

    initSideBar()
    initGroupPanel()

    gridView.setCellWidth(300)
    gridView.setCellHeight(600)

    gridView.setCellFactory { _ =>
      val loader = new SimpleFxmlLoader[HeaderController]("header.fxml")

      new GridCell[Game] {
        itemProperty().addListener { (_, _, g) => {
          setGraphic({
            if (g != null) {
              loader.controller.load(g)
              loader.node
            } else
              null
          })
        }
        }
      }

    }

    popPanel.setAutoHide(true)
    popPanel.setAnimated(false)
  }

  private def initGroupPanel() = {
    cvList.setCellFactory(_ =>
      NodeListCell[DataItem] {
        case SampleItem(title, count) =>
          new HBox(Tags.toNodes(title), new Label(String.valueOf(count)))
      }
    )

    cvList.getSelectionModel.selectedItemProperty().addListener((_, _, cv) => {
      cv match {
        case SampleItem(title, _) =>
          val defaultP: Predicate[Game] = (g: Game) => Option(g.gameCharacters).exists(_.asScala.exists(_.getShowCV().exists(_ == title)))

          groupPredicate = defaultP
          val filterPredicate = filterPanelController.predicate
          val p = if (filterPredicate != null) groupPredicate.and(filterPredicate)
          else groupPredicate
          filteredGames.setPredicate(p)
          recount()
        case _ =>
      }

    })

    tagList.setCellFactory(_ =>
      NodeListCell[DataItem] {
        case SampleItem(title, count) =>
          new HBox(Tags.toNodes(title), new Label(String.valueOf(count)))
      }
    )

    groupCVServ.valueProperty.addListener((_, _, newValue) => {
      if (newValue != null) {
        cvList.getSelectionModel.clearSelection()
        cvList.setItems(FXCollections.observableList(newValue))
      }
    })
    groupTagServ.valueProperty.addListener((_, _, newValue) => {
      if (newValue != null) tagList.setItems(FXCollections.observableList(newValue))
    })
  }


  private def initSideBar() = {

    initSidebarContentView()
  }

  private def initSidebarContentView() = {
    filterPanelController.onSetProperty.addListener((_, _, newV) => {
      if (newV) {
        val load = groupPredicate == null
        val filterPredicate = filterPanelController.predicate
        val p = if (groupPredicate != null) filterPredicate.and(groupPredicate)
        else filterPredicate
        filteredGames.setPredicate(p)
        recount()
        if (load) setSideBarData(filteredGames)
      }

    })
    brandGroupController.onSetProperty.addListener((_, _, newValue) => {
      if (newValue) {
        groupPredicate = brandGroupController.predicate
        val filterPredicate = filterPanelController.predicate
        val p = if (filterPredicate != null) groupPredicate.and(filterPredicate)
        else groupPredicate
        filteredGames.setPredicate(p)
        recount()
      }

    })
    dateGroupController.onSetProperty.addListener((_, _, newValue) => {
      if (newValue) {
        groupPredicate = dateGroupController.predicate
        val filterPredicate = filterPanelController.predicate
        val p = if (filterPredicate != null) groupPredicate.and(filterPredicate)
        else groupPredicate
        filteredGames.setPredicate(p)
        recount()
      }

    })
  }

  def recount(): Unit = {
    tablelist.scrollTo(0)
    resetCount(filteredGames)
  }

  def load(games: ObservableList[Game], initPredicate: Predicate[Game] = null): Unit = {
    filterPanel.setVisible(true)
    groupPredicate = null
    filteredGames = new FilteredList(games)


    val initP = HGameApp.mergeP(initPredicate)
    // set filter
    filteredGames.setPredicate(initP)
    recount()

    // set defaultPredicate
    filterPanelController.predicate = initP
    val sortedData = new SortedList[Game](filteredGames)
    sortedData.comparatorProperty.bind(tablelist.comparatorProperty)
    loadItems(sortedData)
    setSideBarData(filteredGames)
  }


  private def setSideBarData(filteredGames: FilteredList[Game]) = {
    dateGroupController.init(filteredGames)
    brandGroupController.init(filteredGames)
    groupCVServ.restart()
    groupTagServ.restart()
  }

  private def loadItems(sortedData: SortedList[Game]) = {
    tablelist.setItems(sortedData)
    tablelist.scrollTo(0)

    gridView.setItems(sortedData)

    //    loadFlow(sortedData)
  }

  private def resetCount(filteredGames: util.List[Game]) = {

    lbItemCount.setText(s"${filteredGames.size} 件")
    //    loadFlow(filteredGames)

  }

  //  private def loadFlow(filteredGames: util.List[Game]) = {
  //
  //    val iamges = filteredGames.listIterator().asScala.to(LazyList).map { game =>
  //
  //
  //
  //    }.asJava
  //
  //    //    gridView.setItems(filteredGames)
  //
  //  }

}